File: benchmarks/bench_strings.php
<?php

// Concatenate "hello" 500,000 times
$s = "";
for ($i = 0; $i < 500000; $i++) {
    $s .= "hello";
}
echo "concat len: " . strlen($s) . "\n";

// Split a large string 100,000 times
$csv = "alpha,bravo,charlie,delta,echo,foxtrot,golf,hotel";
$total_parts = 0;
for ($j = 0; $j < 100000; $j++) {
    $parts = explode(",", $csv);
    $total_parts += count($parts);
}
echo "split parts: $total_parts\n";

// Regex replace on a template string 200,000 times
$template = "Hello NAME, welcome to PLACE on DATE";
$result = "";
for ($m = 0; $m < 200000; $m++) {
    $result = $template;
    $result = preg_replace("/NAME/", "World", $result, 1);
    $result = preg_replace("/PLACE/", "Strada", $result, 1);
    $result = preg_replace("/DATE/", "today", $result, 1);
}
echo "regex result: $result\n";

File: benchmarks/bench_strings.go
// Go counterpart of bench_strings.strada. The accumulating concat uses
// strings.Builder (the idiomatic Go equivalent of JS cons-string concat);
// the regex passes use the stdlib regexp package, first-match only to
// mirror JS's non-global .replace().
package main

import (
	"fmt"
	"regexp"
	"strings"
)

func replaceFirst(re *regexp.Regexp, s, repl string) string {
	loc := re.FindStringIndex(s)
	if loc == nil {
		return s
	}
	return s[:loc[0]] + repl + s[loc[1]:]
}

func main() {
	var b strings.Builder
	for i := 0; i < 500000; i++ {
		b.WriteString("hello")
	}
	s := b.String()
	fmt.Printf("concat len: %d\n", len(s))

	csv := "alpha,bravo,charlie,delta,echo,foxtrot,golf,hotel"
	totalParts := 0
	for j := 0; j < 100000; j++ {
		parts := strings.Split(csv, ",")
		totalParts += len(parts)
	}
	fmt.Printf("split parts: %d\n", totalParts)

	template := "Hello NAME, welcome to PLACE on DATE"
	patName := regexp.MustCompile("NAME")
	patPlace := regexp.MustCompile("PLACE")
	patDate := regexp.MustCompile("DATE")
	result := ""
	for m := 0; m < 200000; m++ {
		result = template
		result = replaceFirst(patName, result, "World")
		result = replaceFirst(patPlace, result, "Strada")
		result = replaceFirst(patDate, result, "today")
	}
	fmt.Printf("regex result: %s\n", result)
}
